### ISSUE TITLE
vm.getPrimaryNetworkInterface().primaryIPConfiguration().inner().subnet().networkSecurityGroup() returns null

### ISSUE THREAD

**Issue by barryku** [Azure/azure-sdk-for-java#1794]
We want to display security group ID along with other VM data, but getting it through VM returns null.  The current workaround is to get list of subnets in the same resource group, and then use the subnet object to get the data we need.  However, it takes twice as much time to return the data then before due to the extra query for subnets.  Both the following return the security group properly from subnet object,

```
subnet.inner().networkSecurityGroup();
subnet.getNetworkSecurityGroup();
```
I am pretty sure I am getting the right subnet since vm.getPrimaryNetworkInterface().primaryIPConfiguration().inner().subnet().id() has the same id of my subnet object that's coming from azure.networks().listByResourceGroup(resourceGroupName);

**Comment by martinsawicki** [comment]
That's because the code sample shown in the title of this issue goes though `.inner()` to get to the Subnet. The auto-gen'd "inner" objects (in this case `SubnetInner`), when returned as references from other "inner" objects (in this case `NetworkInterfaceIPConfigurationInner`), do not contain the entirety of the referenced object - only the ID field is reliable.

Overall, it is not recommended to reference anything under `.inner()` directly since that's where you get into the auto-generated and under-documented API territory, whose primary goal is to be used by the SDK implementation itself under the hood, and not so much by user code. The `.inner()` method is exposed to users only for advanced scenarios as a temporary workaround whenever it turns out that some functionality exposed by that auto-generated API is not yet covered by the outer, "hand-coded", user-facing fluent layer. But one really needs to know the inner-workings of that auto-generated implementation layer to be able to successfully navigate its API.

So instead, to get to the NSG from a VM, what is needed first is the info about the Subnet, so that it can be fetched as a proper object, via the outer, user-facing APIs only:

```java
NetworkInterface primaryNic = vm.getPrimaryNetworkInterface();
NicIPConfiguration primaryIpConfig = primaryNic.primaryIPConfiguration();
Network network = primaryIpConfig.getNetwork();
Subnet subnet = network.subnets().get(primaryIpConfig.subnetName());
```

At this point we have a properly constructed reference to a `Subnet` object, so to look up its NSG, we just do this:

```java
NetworkSecurityGroup nsg = subnet.getNetworkSecurityGroup();
```

(if all you need is the ID of the NSG, then that's also available at that stage: `subnet.networkSecurityGroupId()`)

So this is how you would fetch an NSG starting from a VM today (as of v1.0+). I've added a test case that goes through this scenario [here](https://github.com/Azure/azure-sdk-for-java/blob/master/azure-mgmt-compute/src/test/java/com/microsoft/azure/management/compute/VirtualMachineOperationsTests.java#L96).

Now that this question has been asked though, I do recognize that some extra convenience could be introduced here to make this scenario less verbose. So in the upcoming v1.2, this logic will get a little bit shorter:

```java
NetworkInterface primaryNic = vm.getPrimaryNetworkInterface();
NicIPConfiguration primaryIpConfig = primaryNic.primaryIPConfiguration();
NetworkSecurityGroup nsg = primaryIpConfig.getNetworkSecurityGroup();
```

as shown in the aforementioned test case as well.  The `NicIPConfiguration.getNetworkSecurityGroup()` is the enhancement here.

Note however that the multiple calls to Azure happening here cannot really be avoided. The Subnet info is not available until the Subnet is explicitly fetched, as a separate call to Azure.

**Comment by martinsawicki** [comment]
explanation provided, extra convenience methods released in v1.2, so closing.
